File: src/lib/elaya/provider.ts
// Elaya LLM provider contract — THE provider-neutral boundary.
//
// Every shape in this file is provider-agnostic. Anthropic/Gemini/OpenAI request
// and response formats are normalized INSIDE each adapter (lib/elaya/adapters/*)
// and must never leak past it — the brain, tools, services and UI only ever see
// these types. Adding a provider means adding one adapter file and one row in
// llm_providers; nothing else changes.

import type { ElayaToolCallRecord } from '@/lib/types/elaya';

/**
 * A file shown to the model inside a user turn — an image or a PDF, as bytes.
 *
 * Provider-neutral by construction: base64 + a media type is the shape every
 * major provider accepts (Anthropic image/document blocks, Gemini inlineData,
 * OpenAI image_url with a data: URL). An adapter that cannot show files should
 * drop them rather than throw — the text of the turn still stands on its own.
 *
 * The bytes are held in memory for the length of the call and never stored:
 * the file itself already lives wherever it came from (the freshdesk-attachments
 * bucket, say) and the model sees a copy, not a new home for it.
 */
export type LlmFilePart = {
  /** 'image/png', 'image/jpeg', 'image/gif', 'image/webp', or 'application/pdf'. */
  mediaType: string;
  /** The file's bytes, base64-encoded, with no data: prefix. */
  dataBase64: string;
};

/**
 * One turn in the model conversation, provider-neutral.
 *
 * A user turn may carry `files` alongside its text (added 2026-09-17 for the
 * vendor extractor, which reads invoices that were photographed rather than
 * typed — about 39% of concierge notes carry a file and the supplier's details
 * are commonly ONLY in it). Text stays required: a file with no instruction is
 * not a question, and every existing caller keeps working unchanged.
 */
export type LlmChatMessage =
  | { role: 'user'; content: string; files?: LlmFilePart[] }
  | { role: 'assistant'; content: string; toolCalls?: ElayaToolCallRecord[] }
  | { role: 'tool'; toolCallId: string; content: string };

/** JSON-Schema tool definition (the lowest common denominator across providers). */
export type LlmToolDefinition = {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
};

export type LlmStopReason = 'end_turn' | 'tool_use' | 'max_tokens' | 'refusal' | 'other';

export type LlmUsage = {
  inputTokens: number;
  outputTokens: number;
};

export type LlmCompleteRequest = {
  model: string;
  maxTokens: number;
  system: string;
  messages: LlmChatMessage[];
  tools?: LlmToolDefinition[];
  /**
   * When true, the adapter SHOULD mark the stable prefix (tools + system) with a
   * provider-native prompt-cache breakpoint so calls 2..n of a multi-tool turn
   * read it at ~0.1x instead of re-billing it at full input rate. Provider-neutral
   * by design: an adapter without caching simply ignores it (correctness is
   * identical either way — caching changes billing only, never what the model
   * sees). The CALLER guarantees the prefix is byte-stable across the turn (no
   * per-request timestamp/UUID in `system` or `tools`) — see persona.ts
   * buildElayaTimeContext, which keeps the volatile "today" anchor OUT of `system`.
   */
  cachePrefix?: boolean;
  /** Streamed text deltas (assistant prose only — never tool-call JSON). */
  onTextDelta?: (delta: string) => void;
};

export type LlmCompleteResult = {
  /** Full assistant prose for the turn (concatenation of streamed deltas). */
  text: string;
  /** Normalized tool calls requested this turn (empty when none). */
  toolCalls: ElayaToolCallRecord[];
  stopReason: LlmStopReason;
  usage: LlmUsage;
};

/**
 * THE one contract every adapter implements. Streaming is part of complete():
 * adapters stream internally and emit deltas through onTextDelta, then resolve
 * with the normalized final result.
 */
export interface LlmProviderAdapter {
  readonly name: string;
  complete(req: LlmCompleteRequest): Promise<LlmCompleteResult>;
}

File: backend/app/llm/provider.py
"""The provider-neutral LLM contract — the Python port of lib/elaya/provider.ts.

Every shape here is provider-agnostic. Anthropic (and later Google/OpenAI)
request/response formats are normalized INSIDE each adapter and never leak
past it: the router, the loop, and the tools only ever see these types.
Adding a provider = one adapter module + one llm_providers row.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import Any, Literal

StopReason = Literal["end_turn", "tool_use", "max_tokens", "refusal", "other"]


@dataclass
class ToolCall:
    """One normalized tool invocation requested by the model."""

    id: str
    name: str
    input: dict[str, Any]


@dataclass
class ChatMessage:
    """One provider-neutral conversation turn.

    role 'tool' carries a tool RESULT back to the model (tool_call_id set).
    """

    role: Literal["user", "assistant", "tool"]
    content: str
    tool_calls: list[ToolCall] = field(default_factory=list)
    tool_call_id: str | None = None
    # The provider's OWN content blocks for an assistant turn, kept verbatim so the next
    # request in the same loop replays them unchanged (thinking blocks with their
    # signatures, and the tool-search blocks that make a discovered tool stay loaded).
    # Opaque to everything outside the adapter; None for persisted history, which
    # replays as text only.
    raw_blocks: list[dict[str, Any]] | None = None


@dataclass
class ToolDefinition:
    name: str
    description: str
    input_schema: dict[str, Any]
    # A deferred tool is in the catalog but not in the model's context until the model
    # searches for it (the 2026 answer to "too many tools": the hot set stays small and
    # sharp, the long tail stays one search away). False = loaded up front.
    defer_loading: bool = False


@dataclass
class CompleteRequest:
    model: str
    max_tokens: int
    system: str
    messages: list[ChatMessage]
    tools: list[ToolDefinition] = field(default_factory=list)
    # When True the adapter marks the stable prefix (tools + system) with a
    # provider-native prompt-cache breakpoint: calls 2..n of a multi-tool turn
    # re-read it at ~0.1x. The CALLER keeps the prefix byte-stable across the
    # turn (no timestamps/UUIDs in system or tools) — same contract as the
    # Node provider.ts.
    cache_prefix: bool = False
    # A VOLATILE trailing system block (the per-turn time anchor) delivered
    # AFTER the cache_control breakpoint — it changes every request without
    # busting the cached prefix. None = no tail.
    system_tail: str | None = None
    on_text_delta: Callable[[str], Awaitable[None]] | None = None
    # Give the model the provider's tool-search tool so it can discover deferred tools by
    # describing what it needs. The adapter adds the provider-native search tool; the
    # loop turns this on whenever any tool in `tools` is deferred.
    tool_search: bool = False


@dataclass
class CompleteResult:
    text: str
    tool_calls: list[ToolCall]
    stop_reason: StopReason
    input_tokens: int
    output_tokens: int
    # The assistant turn's content blocks as the provider returned them (see
    # ChatMessage.raw_blocks). Empty when the adapter has nothing to preserve.
    raw_content: list[dict[str, Any]] = field(default_factory=list)
